### PR TITLE
Purchases

### PR DESCRIPTION
**Pull request by steviekaye**

### CODE REVIEW

File: app/controllers/purchases_controller.rb [hunk 1]
```diff
@@ -40,6 +40,6 @@ def category
   private
 
   def purchase_params
-    params.require(:purchase).permit(:date, :description, :amount, :payee, :subcategory, :user)
+    params.require(:purchase).permit(:date, :description, :amount, :payee, :subcategory)
```
Comment (kfrn) [inline app/controllers/purchases_controller.rb]: Should this also include `:purchaser` and `:purchasee`?

File: app/controllers/purchases_controller.rb [hunk 2]
```diff
@@ -14,9 +14,9 @@ def edit; end
   def create
     @categories = Category.all
     @purchase = Purchase.new(purchase_params)
-
     @purchase.subcategory = Subcategory.find(params[:subcategory_id]) if params[:subcategory_id].present?
-    @purchase.user = User.find(params[:user_id]) if params[:user_id].present?
+    @purchase.purchaser = User.find(params[:purchase][:purchaser_id]) if params[:purchase][:purchaser_id].present?
+    @purchase.purchasee = User.find(params[:purchase][:purchasee_id]) if params[:purchase][:purchasee_id].present?
```
Comment (kfrn) [inline app/controllers/purchases_controller.rb]: I'm just wondering about the `.present?` calls here - I don't know it offhand, but it seems like there should be some way to get around this.

File: app/models/purchase.rb [hunk 3]
```diff
@@ -1,13 +1,13 @@
 class Purchase < ApplicationRecord
-  belongs_to :user
   belongs_to :subcategory
+  belongs_to :purchaser, class_name: 'User', foreign_key: 'purchaser_id'
+  belongs_to :purchasee, class_name: 'User', foreign_key: 'purchasee_id'
```
Comment (kfrn) [inline app/models/purchase.rb]: I wonder if `accepts_nested_attributes_for` might be useful here? (Disclaimer, I haven't used it myself).
* [Rails docs](https://api.rubyonrails.org/classes/ActiveRecord/NestedAttributes/ClassMethods.html)
* [Article](https://medium.com/@onyoo/why-accepts-nested-attributes-for-6ed190def58a)

File: app/views/purchases/_form.html.erb [hunk 4]
```diff
@@ -51,18 +51,28 @@
...
+</div>
+
+<label class="label">Purchasee</label>
+<div class="field is-grouped">
+
+<%= form.collection_radio_buttons( :purchasee_id, User.all, :id, :name) do |b| %>
+<div class="control">
+  <%= b.radio_button %>
+  <%= b.label %>
+</div>
```
Comment (kfrn) [inline app/views/purchases/_form.html.erb]: See note above about `accepts_nested_attributes_for` - if you use that, there are implications for the view.

File: app/views/purchases/index.html.erb [hunk 5]
```diff
@@ -20,19 +20,22 @@
           <th>Payee</th>
           <th>Category</th>
           <th>Purchaser</th>
+          <th>Purchasee</th>
         </tr>
       </thead>
       <tbody>
 
-        <% @purchases.each do |purchase| %>
+        <% @purchases.order(:date).reverse_order.each do |purchase| %>
```
Comment (kfrn) [inline app/views/purchases/index.html.erb]: I think this slipped by in your merge conflict resolution - you changed the ordering in the controller now :)

File: app/views/purchases/index.html.erb [hunk 6]
```diff
@@ -20,19 +20,22 @@
...
           <tr>
             <td><%= purchase.date.to_s %></td>
             <td><%= purchase.description %></td>
             <td><%= number_to_currency(purchase.amount) %></td>
             <td><%= purchase.payee %></td>
             <td><%= Subcategory.find(purchase[:subcategory_id]).name %></td>
-            <td><%= User.find(purchase[:user_id]).name %></td>
+            <td><%= User.find(purchase[:purchaser_id]).name %></td>
+            <td><%= User.find(purchase[:purchasee_id]).name %></td>
```
Comment (kfrn) [inline app/views/purchases/index.html.erb]: I'm starting to wonder if it'd be better to set these as instance variables in the controller, e.g., `@subcategory`, `@purchaser`, `@purchasee`. Might pay to look up if there's a convention for this kind of thing.

Comment (steviekaye) [reply app/views/purchases/index.html.erb]: Scratch the above, but should be able to use helpers like purchase.purchasee.name in place of User.find(purchase[:purchasee_id]).name

File: app/controllers/purchases_controller.rb (hunk 1 above)
Comment (steviekaye) [reply app/controllers/purchases_controller.rb]: I believe those cases are covered/caught by the 'belongs_to :purchaser ...' lines in the model, so adding them at this point would be redundant - what's the best practice?

File: app/views/purchases/index.html.erb (hunk 5 above)
Comment (steviekaye) [reply app/views/purchases/index.html.erb]: Nice catch!

File: app/views/purchases/index.html.erb (hunk 6 above)
Comment (steviekaye) [reply app/views/purchases/index.html.erb]: Great suggestion - much tidier to look at and fewer database calls.

**Pull request merged**